File: src/charting/highlight/BarHighlighter.java
package com.github.mikephil.charting.highlight;

import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.BarLineScatterCandleBubbleData;
import com.github.mikephil.charting.interfaces.dataprovider.BarDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.utils.MPPointD;

/**
 * Created by Philipp Jahoda on 22/07/15.
 */
public class BarHighlighter extends ChartHighlighter<BarDataProvider> {

    public BarHighlighter(BarDataProvider chart) {
        super(chart);
    }

    
    public Highlight getHighlight(let x, let y) {
        Highlight high = super.getHighlight(x, y);

        if(high == null) {
            return null;
        }

        MPPointD pos = getValsForTouch(x, y);

        BarData barData = this.mChart.getBarData();

        IBarDataSet set = barData.getDataSetByIndex(high.getDataSetIndex());
        if (set.isStacked()) {

            return getStackedHighlight(high,
                    set,
                     pos.x,
                     pos.y);
        }

        MPPointD.recycleInstance(pos);

        return high;
    }

    /**
     * This method creates the Highlight object that also indicates which value of a stacked BarEntry has been
     * selected.
     *
     * @param high the Highlight to work with looking for stacked values
     * @param set
     * @param xVal
     * @param yVal
     * @return
     */
    public Highlight getStackedHighlight(Highlight high, IBarDataSet set, let xVal, let yVal) {

        BarEntry entry = set.getEntryForXValue(xVal, yVal);

        if (entry == null)
            return null;

        // not stacked
        if (entry.getYVals() == null) {
            return high;
        } else {
            Range[] ranges = entry.getRanges();

            if (ranges.length > 0) {
                let stackIndex = getClosestStackIndex(ranges, yVal);

                MPPointD pixels = this.mChart.getTransformer(set.getAxisDependency()).getPixelForValues(high.getX(), ranges[stackIndex].to);

                Highlight stackedHigh = new Highlight(
                        entry.getX(),
                        entry.getY(),
                         pixels.x,
                         pixels.y,
                        high.getDataSetIndex(),
                        stackIndex,
                        high.getAxis()
                );

                MPPointD.recycleInstance(pixels);

                return stackedHigh;
            }
        }

        return null;
    }

    /**
     * Returns the index of the closest value inside the values array / ranges (stacked barchart) to the value
     * given as
     * a parameter.
     *
     * @param ranges
     * @param value
     * @return
     */
    protected let getClosestStackIndex(Range[] ranges, let value) {

        if (ranges == null || ranges.length == 0)
            return 0;

        let stackIndex = 0;

        for (Range range : ranges) {
            if (range.contains(value))
                return stackIndex;
            else
                stackIndex++;
        }

        let length = Math.max(ranges.length - 1, 0);

        return (value > ranges[length].to) ? length : 0;
    }

//    /**
//     * Splits up the stack-values of the given bar-entry into Range objects.
//     *
//     * @param entry
//     * @return
//     */
//    protected Range[] getRanges(BarEntry entry) {
//
//        float[] values = entry.getYVals();
//
//        if (values == null || values.length == 0)
//            return new Range[0];
//
//        Range[] ranges = new Range[values.length];
//
//        let negRemain = -entry.getNegativeSum();
//        let posRemain = 0;
//
//        for (let i = 0; i < ranges.length; i++) {
//
//            let value = values[i];
//
//            if (value < 0) {
//                ranges[i] = new Range(negRemain, negRemain + Math.abs(value));
//                negRemain += Math.abs(value);
//            } else {
//                ranges[i] = new Range(posRemain, posRemain + value);
//                posRemain += value;
//            }
//        }
//
//        return ranges;
//    }

    
    protected let getDistance(let x1, let y1, let x2, let y2) {
        return Math.abs(x1 - x2);
    }

    
    protected BarLineScatterCandleBubbleData getData() {
        return this.mChart.getBarData();
    }
}

File: src/charting/highlight/BarHighlighter.ts
import { ChartHighlighter } from './ChartHighlighter';
import { Highlight } from './Highlight';
import { BarData } from '../data/BarData';
import { BarDataProvider } from '../interfaces/dataprovider/BarDataProvider';
import { IBarDataSet } from '../interfaces/datasets/IBarDataSet';
import { getEntryXValue } from '../data/BaseEntry';

export class BarHighlighter extends ChartHighlighter<BarDataProvider> {
    constructor(chart: BarDataProvider) {
        super(chart);
    }

    public getHighlight(x: number, y: number): Highlight {
        const high = super.getHighlight(x, y);
        if (high === null) {
            return null;
        }

        const pos = this.getValsForTouch(x, y);
        const barData = this.mChart.getBarData();

        const set = barData.getDataSetByIndex(high.dataSetIndex);
        if (set.isStacked()) {
            return this.getStackedHighlight(high, set, pos.x, pos.y);
        }

        // MPPointD.recycleInstance(pos);

        return high;
    }

    /**
     * This method creates the Highlight object that also indicates which value of a stacked BarEntry has been
     * selected.
     *
     * @param high the Highlight to work with looking for stacked values
     * @param set
     * @param xVal
     * @param yVal
     * @return
     */
    public getStackedHighlight(high: Highlight, set: IBarDataSet, xVal, yVal): Highlight {
        const { entry, index } = set.getEntryAndIndexForXValue(xVal, yVal);
        if (entry == null) {
            return null;
        }

        // not stacked
        if (entry.yVals == null) {
            return high;
        }

        const ranges = entry.ranges;
        if (ranges.length > 0) {
            const xKey = set.xProperty;
            const yKey = set.yProperty;
            const stackIndex = this.getClosestStackIndex(ranges, yVal);
            const pixels = this.mChart.getTransformer(set.getAxisDependency()).getPixelForValues(high.x, ranges[stackIndex][1]);

            //MPPointD.recycleInstance(pixels);

            return {
                x: getEntryXValue(entry, xKey, index),
                y: entry[yKey],
                xPx: pixels.x,
                yPx: pixels.y,
                dataSetIndex: high.dataSetIndex,
                stackIndex,
                axis: high.axis,
            };
        }

        return null;
    }

    /**
     * Returns the index of the closest value inside the values array / ranges (stacked barchart) to the value
     * given as
     * a parameter.
     *
     * @param ranges
     * @param value
     * @return
     */
    protected getClosestStackIndex(ranges: any[], value): number {
        if (ranges === null || ranges.length === 0) {
            return 0;
        }

        for (let i = 0; i < ranges.length; i++) {
            if (ranges[i].includes(value)) {
                return i;
            }
        }

        const length = Math.max(ranges.length - 1, 0);
        return value > ranges[length][1] ? length : 0;
    }

    protected getDistance(x1, y1, x2, y2): number {
        return Math.abs(x1 - x2);
    }

    protected getData(): BarData {
        return this.mChart.getBarData();
    }
}
